src/sidebar/services/auth.ts

```diff
@@ -0,0 +1,335 @@
+    this._tokenInfoPromise = null;
+
+    this._apiRoutes = apiRoutes;
+    this._localStorage = localStorage;
+    this._settings = settings;
+    this._toastMessenger = toastMessenger;
+    this._window = $window;
+
+    this._listenForTokenStorageEvents();
+  }
+
+  /**
+   * Listen for updated access and refresh tokens saved by other instances of
+   * the client.
+   */
+  private _listenForTokenStorageEvents() {
+    this._window.addEventListener('storage', ({ key }) => {
+      if (key === this._storageKey()) {
+        // Reset cached token information. Tokens will be reloaded from storage
+        // on the next call to `getAccessToken()`.
+        this._tokenInfoPromise = null;
+        this.emit('oauthTokensChanged');
+      }
+    });
+  }
+
+  /**
+   * Fetch the last-saved access/refresh tokens for `authority` from local
+   * storage.
+   */
+  private _loadToken() {
+    const token = this._localStorage.getObject(this._storageKey());
+
+    if (
+      !token ||
+      typeof token.accessToken !== 'string' ||
+      typeof token.refreshToken !== 'string' ||
+      typeof token.expiresAt !== 'number'
+    ) {
+      return null;
+    }
+
+    return {
+      accessToken: token.accessToken,
+      refreshToken: token.refreshToken,
+      expiresAt: token.expiresAt,
+    };
+  }
+
+  /**
+   * Return a cached OAuthClient if available, or configure one and return it.
+   */
+  async _oauthClient() {
+    if (this._client) {
+      return this._client;
+    }
+
+    const links = await this._apiRoutes.links();
+    this._client = new OAuthClient({
+      clientId: this._settings.oauthClientId,
+      authorizationEndpoint: links['oauth.authorize'],
+      revokeEndpoint: links['oauth.revoke'],
+      tokenEndpoint: resolve('token', this._settings.apiUrl),
+    });
+    return this._client;
+  }
+
+  /**
+   * Exchange the authorization code retrieved from login popup for a new
+   * access token.
+   */
+  private async _exchangeAuthCodeForToken(code: string) {
+    const client = await this._oauthClient();
+    const tokenInfo = await client.exchangeAuthCode(code);
+    this._saveToken(tokenInfo);
+    return tokenInfo;
+  }
+
+  /**
+   * Attempt to exchange a grant token for an access token.
+   *
+   * Publishers or other embedders can set a grant token for a user in service
+   * configuration.
+   */
+  private async _exchangeGrantToken(grantToken: string) {
+    if (grantToken) {
+      const client = await this._oauthClient();
+      // User is logged-in on the publisher's website.
+      // Exchange the grant token for a new access token.
+      return client.exchangeGrantToken(grantToken).catch(err => {
+        this._toastMessenger.error(
+          `Hypothesis login lost: You must reload the page to annotate.`,
+          {
+            autoDismiss: false,
+          }
+        );
+        throw err;
+      });
+    } else {
+      // Grant token is empty: user is anonymous on the publisher's website.
```

Comment:
Is the grant token actually an empty string in this case or some other nullish value? If so we'll need to update the signature of this function.

Reply:
I'm presuming (as I inherited this code) that it's the empty string. I'll go ensure that's how this configuration value is coerced if it's not present.

Reply:
After a bit of reflection, I've refactored this a bit to avoid confusion or potential future bugs. The checks for the cases in which `grantToken` is undefined (no provided `serviceConfig` auth info) or present-but-empty (user is not auth'd on publisher/embedder site) are both handled in the flow of `getAccessToken`, and `_exchangeGrantToken` now expects a (non-empty) string.